src/mito2/src/region.rs

```diff
@@ -279,12 +292,133 @@ impl MitoRegion {
         )
     }
 
-    /// Sets the region to readonly gracefully. This acquires the manifest write lock.
-    pub(crate) async fn set_role_state_gracefully(&self, state: SettableRegionRoleState) {
-        let _manager = self.manifest_ctx.manifest_manager.write().await;
-        // We acquires the write lock of the manifest manager to ensure that no one is updating the manifest.
-        // Then we change the state.
-        self.set_role(state.into());
+    /// Sets the staging state.
+    /// You should call this method in the worker loop.
+    /// Transitions from Writable to Staging state.
+    pub(crate) fn set_staging(&self) -> Result<()> {
+        self.compare_exchange_state(
+            RegionLeaderState::Writable,
+            RegionRoleState::Leader(RegionLeaderState::Staging),
+        )
+    }
+
+    /// Exits the staging state back to writable.
+    /// You should call this method in the worker loop.
+    /// Transitions from Staging to Writable state.
+    pub(crate) fn exit_staging(&self) -> Result<()> {
+        self.compare_exchange_state(
+            RegionLeaderState::Staging,
+            RegionRoleState::Leader(RegionLeaderState::Writable),
+        )
+    }
+
+    /// Sets the region role state gracefully. This acquires the manifest write lock.
+    pub(crate) async fn set_role_state_gracefully(
+        &self,
+        state: SettableRegionRoleState,
+    ) -> Result<()> {
+        let current_state = self.state();
```

Comment:
This method no longer takes the manifest lock and may break existing assumptions. @WenyXu What do you think?
We may still have to take the manifest lock.